### ISSUE TITLE
removing features

### ISSUE THREAD

**Issue by gkz** [gkz/LiveScript#219]
So far (on master), slated for removal (and currently with compiler notices upon use):
- short function syntax - `f(x) = x` - [currently has notice]
  why: hacky implementation, duplicating functionaly
  use --> to make curried functions
- `+++` concat op [currently has notice]
  why: added ++ as concat op, which is shorter and better
- `where` statement - relatively useless - just use local var assignment or a let statement  [currently has notice]
- `undefined` alias for `void` - added for CoffeeScript compatibility, but with coffee2ls that is not an issue  [currently has notice]
- `!?` inexistence op  [currently has notice]

---

discussion:
- choose one of `**` or `^` for power op?
- remove `import` and `importAll`?
- remove `x is not y` to mean `x isnt y`, have it mean `x is !y` instead
- `yes/no/on/off` aliases - added for CoffeeScript compatibility, but now with coffee2ls, this is not an issue

More controversial:
- remove standalone use of `@`
- remove standalone or hanging use of `::`

**Comment by vendethiel** [comment]
remove unjoined strings

**Comment by igl** [comment]
I never use `@` and if decorators ever become a thing in JS we have a clash.
I did not have the pleasure of coding LS in a team but in coffeescribble and I found not just one bug caused by heavy usage of `?`. That disqualifies the whole thing as a safe-guard to me.

I generally prefer not using symbols if i don't have to... Not all symbols in LS have such quirky and placement specific behavior. I still prefer not to be clever just for saving a few characters.

**Comment by kay-is** [comment]
I use `@` a lot. You're right with the readability and decorators aspect (even if I don't think LS has to mirror every feature of JS).
But it's a very simple substitute and for me it feels fast to skim over code and see all the `@attribute`s that are used even without special highlighting.

**Comment by impinball** [comment]
@igl There's no need for decorators in LiveScript outside of changing property descriptors. You can use plain functions for nearly every use case. (I've used this in my own pet projects before)

```ls
inc = (f) -> ->
  @index += 1
  f.apply @, &

class Foo
  -> @index = 2
  foo: inc -> console.log "#{@index}"

wrap = (f) -> -> new List f.apply @, &

class List
  (@value) ->
  map: wrap (f) -> [f i for i in @value]
  filter: wrap (f) -> [i for i in @value | f i]
  each: (f) -> [f i for i in @value]; @
```

**Comment by impinball** [comment]
And things that I wouldn't mind if they were removed:

- `case`, `when`, `then`, `else`, and `default` in switch-case/`match` expressions
  - IMHO, this is where CoffeeScript starts looking not so pretty.

- `yes`, `no`, `on`, and `off` CoffeeScript boolean synonyms.
  - I rarely see these outside of CoffeeScript and YAML.

- `x is not y` -> change to be equivalent to `(x) == (!y)` (principle of least surprise)
  - Python's version is exclusively used for references, usually for objects or types.

    ```py
    type(foo) is not str
    objA is not objB
    x == y
    ```

  - CoffeeScript itself compiles `x is not y` to `x === !y`.

- `undefined`
  - Exclusively use `void`.
  - Clearer that it's not just JavaScript's problematic (in sloppy mode) `undefined`.

- One of `^` or `**`

- `.` alias of the `<<` composition operator (consistency)

- `* 'str'`, `/ 'str'`
  - Use `String::join`, `String::split`, Prelude.ls's `join` function, or its `split` function.
  - Easier to know at a glance.

- `for ever` alias of `loop`
  - Don't need both.
  - Most languages with this construct use `loop`.

**Comment by heavyk** [comment]
great list ...

I've never tried `is not` and always used `isnt` however, now that I think about it, I could see why that would be logical. without thinking about it, I'd write `if x is !y`

I will admit I've used `* 'str'` a lot, so it'd be the only one I miss, but not that much.

so yeah, I'm totally +1 for all that.

**Comment by impinball** [comment]
@heavyk I've used that quite a bit, too, but it tends to throw me a little at first glance, which isn't nice when you're reading code. That's why I put it on there.

**Comment by vendethiel** [comment]
> * 'str', / 'str'
> 
> Use String::join, String::split, Prelude.ls's join function, or its split function.
> Easier to know at a glance.

Literal overloading in general. It'd make sense to have such feature with a type system (and overloadable operators...), but for us right now...

**Comment by impinball** [comment]
@vendethiel +1

**Comment by heavyk** [comment]
@impinball yea I agree. I had a phase where I wrote a bunch of console apps, and I used that operator all the time. I look at that code now, it's like 'wat??'

**Comment by homam** [comment]
I only disagree with 

> `.` alias of the `<<` composition operator (consistency)

**Comment by heavyk** [comment]
I never used that, what does it do?

**Comment by raine** [comment]
Function composition from right to left.

```ls
ls> (+ 1) . (* 2) <| 1
3
```

**Comment by heavyk** [comment]
oh, I didn't know you could do that!

I just looked at your profile, then at the ramda-cli you're working on, and it gave me a good idea. why do we pass options objects around at all? why not pass around option thunks?

every part of anything we write could be composed. I'll give an example...

resizeImage({
  width: 1024,
  height: 768,
  rotate: 0, normalize: ... etc. // these are options
}

so I pass in my options object with values, but if I were to instead pass thunks, I could get two objects that can always change, but depend on each oher. one obj would have the thunks to call of the other object, and vice versa.. objects would turn out the properties of other objects. it looks in my head like water. I need a better way to express this. I will work on this more...

totally had no idea about those function compositions. whole new world just opened... omg

**Comment by impinball** [comment]
@homam To be fair, I have used it before. I have run into a gotcha with it, though: partially applying it. I have found a few times where I had to use `(<<)` because `( . )` always represents property access. You can still do `(foo .)` though.

@heavyk

Here's a use case for function composition: making a lazy chain:

```ls
a = -> [.. for &]
chain-sync = a >> reduce (>>)

read-rc = chain-sync do
  ('.'+)
  (fs.read-file-sync _, 'utf8')
  JSON.parse
```

I've also used a similar trick in a Gulpfile:

```ls
require! { /* ... */ }
a = -> [.. for &]
pipe = a >> reduce (a, b) -> a.pipe b

gulp.task \build ->
  pipe do
    * gulp.src 'src/**/*.ls'
    * sourcemaps.init!
    * livescript!
    * soucemaps.write!
    * gulp.dest 'dest'
```

Doing the same with promises is easy

```ls

**Comment by impinball** [comment]
@raine I use `match` quite a bit. I ended up making an equivalent targeting ES7 virtual methods (expect it to be slow), just to simplify my program significantly (mostly written in ES6/7):

```js
export const method = f => function (...args) {
  return f(this, ...args);
};

const each = (xs, f) => {
  for (let i = 0; i < xs.length; i++) {
    f(xs[i], i);
  }
};

export const then = method((x, f) => f(x));

export const match = method((value, pairs) =>
        each(pairs, (p, i) => assertCallable(p[0], `test[${i}]`))::then(() => {
    let def = pairs.pop();
    if (Array.isArray(def)) {
        pairs.push(def);
        return undef;
    } else {
        assertCallable(def, 'def');
    }

    for (let [g, run] of pairs) {
        if (typeof g === 'function' ? g(value) : value === g) return run(value);
    }

    return def(value);
});

value::match([
  // I found the common case to be a function call with `value`, so I designed
  // it with that in mind
  [test, value => doSomething(value)],
  [otherTest, doSomethingElse],
  value => def(value), // optional
]);
```

The last part in LS would be this:

```ls
match value
  | test => doSomething value
  | otherTest => doSomethingElse value
  | otherwise => def value
```

I enjoy `match`

**Comment by igl** [comment]
@impinball I enjoy `match`a lot too (#584 is the only issue i have with it)

I am very curious about how these changes are going to land. Will there be a 1.5 with breakage?
Will features like es6 class and module support be pushed off further to LS2?
Can there be a LS2 pre-release with a lot of breakage so big changes can happen before locking the keywords? Is forking CoffeeScriptRedux still on the roadmap?

**Comment by impinball** [comment]
That LS2 branch/roadmap is years outdated. LiveScript uses its own lexer
and parser now.

As for an actual roadmap, I don't think there's much to add right now. But
FWIW I do feel that removing all these features should constitute a new
major version (I know this project doesn't follow Semver). Even gradually,
this would cause a lot of code breakage, including in the compiler itself.

**Comment by focused** [comment]
Match is a **must have feature**

**Comment by ozra** [comment]
+1 on all except stand alone `@` (mostly because I have _a lot_ of old code using it).

**Comment by isiahmeadows** [comment]
@ozra I agree WRT the original list (except I also want to keep `C:: == C.prototype`).

@focused I agree, and I wish `match` was documented. It does IMHO have odd behavior with certain cases (it's rather ad-hoc in how it matches things), and could stand to be fixed, but it should still stay around.

**Comment by chrisvfritz** [comment]
I'm disappointed [the single-item array not evaluating to an array "feature"](https://github.com/gkz/LiveScript/issues/373) wasn't removed in 1.5 after all. 😕  It's something that actively causes pain through unexpected behavior, rather than just duplicating behavior.
